Performance improvements for referential integrity checking

al-niessner opened the pull request:
## 🗒️ Summary
Multi step process of cleaning up validate to improve performance:
1. Fixed ValidateTarget to be compliant with Java: hashCode() returned different values when equals() returned true.
2. Fixed ValidateTarget.cachedTargets was implemented in a Utility class so very few creators of ValidateTarget used it. Moved cache to ValidateTarget, privatized constructors, used factory methods (build) for consistent use of cache.
3. Fixing Identifier to be compliant with Java: hashCode() returned different values when equals() returned true.
4. Fix all of the loops to use contains() instead. Those loops are a consequence of (3).

## ⚙️ Test Data and/or Report
Automated unit/regression tests below should pass.

## ♻️ Related Issues
Closes #969

Commit:
performance improvement

Step one: removed the tallest tent pole by fixing the ValidationTarget cache. Not everything was being cached. In fact, little to none of it was. Implementing this sped up the final step that counts on this by 10x. The 11000 strong bundle went from 73 seconds to 7.3 seconds. The overall time remains fixed at 1500 seconds. Not sure if the 73 seconds were simply redistributed or if that is just normal variation of an uncertain process (depends on network download times).

nutjob4life commented on src/main/java/gov/nasa/pds/tools/validate/ValidationTarget.java:
```diff
@@ -189,8 +190,7 @@ public String toString() {
     if (knownHashCode == 0) {
-      String combined = location + ":" + (type == null ? "" : type.toString());
-      knownHashCode = combined.hashCode();
+      knownHashCode = location.hashCode();
```
Is there ever a risk that `location.hashCode()` would be `0`, and thus re-compute the hash code?

I guess it doesn't really matter, though … hash codes aren't tremendously expensive

al-niessner replied:
Yes. Not in infinite loop but will recompute 0 every time. I left the knownHashCode for nostalgia. Besides location is a string and I bet Java keeps a static value of it around just like it does the string since they are immutable. Really should just return location.hashCode() and not try to buffer it but I am a nostalgic person.

nutjob4life replied:
Hey, a little nostalgia now and then is just fine 😌

nutjob4life reviewed (approved):
LGTM 👍

Commit:
correct identifier

Commit:
fix up collections with Identifier

Commit:
clean up last bits

TargetExaminer does not need to read the same file over and over. Do it once, collecting the information needed and then store for later. Yes, another cache.

The look in referential integrity was overly complicated. Reduced it to a HashSet.contains() which really speeds things up.
